// @base44/sdk/dist/modules/entities.js
/**
 * Creates the entities module for the Base44 SDK.
 *
 * @param axios - Axios instance
 * @param appId - Application ID
 * @returns Entities module with dynamic entity access
 * @internal
 */
export function createEntitiesModule(axios, appId) {
    // Using Proxy to dynamically handle entity names
    return new Proxy({}, {
        get(target, entityName) {
            // Don't create handlers for internal properties
            if (typeof entityName !== "string" ||
                entityName === "then" ||
                entityName.startsWith("_")) {
                return undefined;
            }
            // Create entity handler
            return createEntityHandler(axios, appId, entityName);
        },
    });
}
/**
 * Creates a handler for a specific entity.
 *
 * @param axios - Axios instance
 * @param appId - Application ID
 * @param entityName - Entity name
 * @returns Entity handler with CRUD methods
 * @internal
 */
function createEntityHandler(axios, appId, entityName) {
    const baseURL = `/apps/${appId}/entities/${entityName}`;
    return {
        // List entities with optional pagination and sorting
        async list(sort, limit, skip, fields) {
            const params = {};
            if (sort)
                params.sort = sort;
            if (limit)
                params.limit = limit;
            if (skip)
                params.skip = skip;
            if (fields)
                params.fields = Array.isArray(fields) ? fields.join(",") : fields;
            return axios.get(baseURL, { params });
        },
        // Filter entities based on query
        async filter(query, sort, limit, skip, fields) {
            const params = {
                q: JSON.stringify(query),
            };
            if (sort)
                params.sort = sort;
            if (limit)
                params.limit = limit;
            if (skip)
                params.skip = skip;
            if (fields)
                params.fields = Array.isArray(fields) ? fields.join(",") : fields;
            return axios.get(baseURL, { params });
        },
        // Get entity by ID
        async get(id) {
            return axios.get(`${baseURL}/${id}`);
        },
        // Create new entity
        async create(data) {
            return axios.post(baseURL, data);
        },
        // Update entity by ID
        async update(id, data) {
            return axios.put(`${baseURL}/${id}`, data);
        },
        // Delete entity by ID
        async delete(id) {
            return axios.delete(`${baseURL}/${id}`);
        },
        // Delete multiple entities based on query
        async deleteMany(query) {
            return axios.delete(baseURL, { data: query });
        },
        // Create multiple entities in a single request
        async bulkCreate(data) {
            return axios.post(`${baseURL}/bulk`, data);
        },
        // Import entities from a file
        async importEntities(file) {
            const formData = new FormData();
            formData.append("file", file, file.name);
            return axios.post(`${baseURL}/import`, formData, {
                headers: {
                    "Content-Type": "multipart/form-data",
                },
            });
        },
    };
}
